test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
 	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(27)));
 }
 
+BOOST_AUTO_TEST_CASE(ufixed_retrieval)
+{
+	char const* sourceCode = R"(
+		contract test {
+			ufixed a = 3.14;
+
+			function f() returns (ufixed) {
+				return a;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(314, 100, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_retrieval)
+{
+	char const* sourceCode = R"(
+		contract test {
+			fixed a = -3.14;
+
+			function f() returns (fixed) {
+				return a;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	bytes const &result = callContractFunction("f()");
+	ABI_CHECK(result, encodeArgs(createFixed(-314, 100, 18)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_echo)
+{
+	char const* sourceCode = R"(
+		contract test {
+			function f(ufixed a) returns (ufixed) {
+				return a;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(ufixed128x18)", [](u256 const& a) -> u256 { return a; }, 0, 100);
+}
+
+
+BOOST_AUTO_TEST_CASE(fixed_echo)
+{
+	char const* sourceCode = R"(
+		contract test {
+			function f(fixed a) returns (fixed) {
+				return a;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(fixed128x18)", [](u256 const& a) -> u256 { return a; }, -50, -1);
+	testContractAgainstCppOnRange("f(fixed128x18)", [](u256 const& a) -> u256 { return a; }, 0, 50);
```

Comment:
These aren't encoding the number as `fixed128x18`, but rather as `fixed256x0` strictly speaking.